src/hotspot/share/nmt/memMapPrinter.cpp

```diff
@@ -0,0 +1,291 @@
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#include "precompiled.hpp"
+
+#include "logging/logAsyncWriter.hpp"
+#include "gc/shared/collectedHeap.hpp"
+#include "memory/allocation.hpp"
+#include "memory/universe.hpp"
+#include "runtime/nonJavaThread.hpp"
+#include "runtime/osThread.hpp"
+#include "runtime/thread.hpp"
+#include "runtime/threadSMR.hpp"
+#include "runtime/vmThread.hpp"
+#include "nmt/memFlagBitmap.hpp"
+#include "nmt/memMapPrinter.hpp"
+#include "nmt/memTracker.hpp"
+#include "nmt/virtualMemoryTracker.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/growableArray.hpp"
+#include "utilities/ostream.hpp"
+
+// Note: throughout this code we will use the term "VMA" for OS system level memory mapping
+
+/// NMT mechanics
+
+// Short, clear, descriptive names for all possible markers. Note that we only expect to see
+// those that have been used with mmap. Flags left out are printed with their nmt flag name.
+#define NMT_FLAGS_DO(f) \
+  /* flag, short, description */ \
+  f(mtGCCardSet,      "CARDTBL", "GC Card table") \
+  f(mtClassShared,    "CDS", "CDS archives") \
+  f(mtClass,          "CLASS", "Class Space") \
+  f(mtCode,           "CODE", "Code Heap") \
+  f(mtGC,             "GC", "GC support data (e.g. bitmaps)") \
+  f(mtInternal,       "INTERN", "Internal") \
+  f(mtJavaHeap,       "JAVAHEAP", "Java Heap") \
+  f(mtOther,          "JDK", "allocated by JDK libraries other than VM") \
+  f(mtMetaspace,      "META", "Metaspace nodes (non-class)") \
+  f(mtSafepoint,      "POLL", "Polling pages") \
+  f(mtThreadStack,    "STACK", "(known) Thread Stack") \
+  f(mtTest,           "TEST", "JVM internal test mappings")
+  //end
+
+static const char* get_shortname_for_nmt_flag(MEMFLAGS f) {
+#define DO(flag, shortname, text) if (flag == f) return shortname;
+  NMT_FLAGS_DO(DO)
+#undef DO
+  return NMTUtil::flag_to_enum_name(f);
+}
+
+/// NMT virtual memory
+
+static bool range_intersects(const void* from1, const void* to1, const void* from2, const void* to2) {
+  const void* const min = MAX2(from1, from2);
+  const void* const max = MIN2(to1, to2);
+  return min < max;
+}
+
+// A Cache that correlates range with MEMFLAG, optimized to be iterated quickly
+// (cache friendly).
+class CachedNMTInformation : public VirtualMemoryWalker {
+  struct Range { const void* from; const void* to; };
+  // Unfortunately, we need to allocate manually, raw, since we must prevent
+  // NMT deadlocks (ThreadCritical).
+  Range* _ranges;
+  MEMFLAGS* _flags;
```

Comment:
Why did you decide to have two separate memory chunks?

I think I'd have used a struct to hold Range* and MEMFLAGS* and use that struct in a single array.

Reply:
Two reasons:

- more condensed, needs less memory. MEMFLAGS is a byte, Range is 16 bytes. A combined structure would be 24 bytes with padding (since the structure needs to be pointer aligned)
- It is more cache friendly, hence faster. I need to iterate through all ranges, but only need to access the MEMFLAGS for the one range I find. By keeping the ranges condensed, more of them fit into a cache line, so iteration is faster.